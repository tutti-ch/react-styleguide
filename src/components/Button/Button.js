import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { filterProps } from "../../helpers/functions";
import Spinner from "../Spinner";
import classes from "./Button.scss";

export default class Button extends Component {
  static LEVEL_PRIMARY = "primary";
  static LEVEL_SECONDARY = "secondary";
  static LEVEL_TERTIARY = "tertiary";

  static ROLE_WARNING = "warning";
  static ROLE_PROMOTE = "promote";

  static SIZE_SMALL = "small";
  static SIZE_MEDIUM = "medium";
  static SIZE_LARGE = "large";

  static POSITION_LEFT = "left";
  static POSITION_RIGHT = "right";
  static POSITION_FULL_WIDTH = "fullWidth";

  static propTypes = {
    /**
     * The content that will be rendered inside the button.
     */
    children: PropTypes.node,

    /**
     * The class name that will be passed to the button.
     */
    className: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.object,
      PropTypes.string
    ]),

    /**
     * The level of the button. This has impact on the shape of the button.
     */
    level: PropTypes.oneOf([
      Button.LEVEL_PRIMARY,
      Button.LEVEL_SECONDARY,
      Button.LEVEL_TERTIARY
    ]).isRequired,

    /**
     * The role of the button. This will change the color of the button.
     */
    role: PropTypes.oneOf([Button.ROLE_PROMOTE, Button.ROLE_WARNING]),

    /**
     * The size of the button.
     */
    size: PropTypes.oneOf([
      Button.SIZE_SMALL,
      Button.SIZE_MEDIUM,
      Button.SIZE_LARGE
    ]).isRequired,

    /**
     * Whether the button is responsive or not. Default is false.
     */
    responsive: PropTypes.bool,

    /**
     * Whether there is an operation going on or not.
     */
    loading: PropTypes.bool,

    /**
     * The position of the button.
     */
    position: PropTypes.oneOf([
      Button.POSITION_LEFT,
      Button.POSITION_RIGHT,
      Button.POSITION_FULL_WIDTH
    ]),

    /**
     * Whether the button is rounded or not
     */
    rounded: PropTypes.bool
  };

  static defaultProps = {
    rounded: true,
    loading: false,
    size: Button.SIZE_MEDIUM,
    type: "submit"
  };

  static contextTypes = {
    onSubmit: PropTypes.func
  };

  state = {
    loading: this.props.loading
  };

  constructor(props, context) {
    super(props);

    const { onSubmit } = context;
    const { type = "submit" } = props;

    if (typeof onSubmit === "function" && type === "submit") {
      onSubmit(({ loading }) => this.setState({ loading }));
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.loading !== this.state.loading) {
      this.setState({ loading: nextProps.loading });
    }
  }

  /**
   * Return the appropriate classes for the button.
   */
  getClasses = () => {
    const {
      level,
      size,
      responsive,
      position,
      role,
      rounded,
      className
    } = this.props;

    return classNames(
      className,
      classes.button,
      classes[level],
      classes[size],
      classes[role],
      classes[position],
      {
        [classes.responsive]: responsive,
        [classes.noRounded]: rounded === false
      }
    );
  };

  /**
   * Render the content for the button.
   */
  renderContent = () => {
    const { children } = this.props;
    const { loading } = this.state;

    if (loading) {
      return [
        <Spinner
          className={classes.spinner}
          size={1.25}
          color={Spinner.COLOR_LIGHT}
          key="spinner"
        />,
        <div className={classes.text} key={"text"}>
          {children}
        </div>
      ];
    }

    return children;
  };

  /**
   * Render the button.
   *
   * @return {*}
   */
  render() {
    const attrs = filterProps(Button.propTypes, this.props);

    // When it is loading, the button is disabled by default.
    if (this.state.loading) {
      attrs.disabled = true;
    }

    return (
      <button className={this.getClasses()} {...attrs}>
        {this.renderContent()}
      </button>
    );
  }
}
