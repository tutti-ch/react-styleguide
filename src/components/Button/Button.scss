@import "../../styles/main";

.button {
  border-radius: $border-radius-s;
  line-height: $size-m;
  border: none;
  padding: 0;
  margin: 0;
  // @include clearfix;
  transition: all $transition-m;
  position: relative;
  text-align: center;
  display: flex;
  align-items: center;
  justify-content: center;

  &:disabled,
  &:disabled:hover {
    background-color: $greyscale-80 !important;
  }

  &.noRounded {
    border-radius: 0;
  }

  :global .svg-sprite {
    width: $size-m;
    height: $size-m;
    margin-right: $size-xxs;
    flex: 0 0 auto;
    align-self: center; // Required for Safari 9.1
  }
}

.left {
  float: left;
  margin-right: $size-m;
}

.right {
  float: right;
  margin-left: $size-m;
}

.responsive {
  width: 100%;

  @include phablet-and-bigger {
    width: auto;
  }
}

.spinner {
  // @include center;
  margin-right: $size-xs;
  line-height: 1;
  display: block;
}

.textLoading {
  visibility: hidden;
}

// SIZE

.button.small {
  line-height: $size-m;
  padding: 0 $size-s;

  height: $size-xl;

  .icon {
    height: $size-xl;
    line-height: $size-m;
  }

  .icon:first-child {
    margin-right: $size-xxs;
  }

  .icon:last-child {
    margin-right: 0;
    margin-left: $size-xxs;
  }

  .icon:only-child {
    margin: 0;
  }
}

.button.medium {
  line-height: $size-m;
  padding: 0 $size-m;

  height: $size-xxl;

  .icon {
    height: $size-xxl;
    line-height: $size-m;
  }

  .icon:first-child {
    margin-right: $size-xs;
  }

  .icon:last-child {
    margin-right: 0;
    margin-left: $size-xs;
  }

  .icon:only-child {
    margin: 0;
  }
}

.button.large {
  line-height: $size-m;
  padding: 0 $size-l;

  height: $size-xxxl;

  .icon {
    height: $size-xxxl;
    line-height: $size-m;
  }

  .icon:first-child {
    margin-right: $size-s;
  }

  .icon:last-child {
    margin-right: 0;
    margin-left: $size-s;
  }

  .icon:only-child {
    margin: 0;
  }
}

// ROLE

.button.primary {
  background-color: $green-50;
  color: $greyscale-100;

  &:hover {
    background-color: $green-60;
  }
}

.button.secondary {
  background-color: $blue-50;
  color: $greyscale-100;

  &:hover {
    background-color: $blue-60;
  }
}

.button.tertiary {
  background-color: $greyscale-100;
  box-shadow: inset 0 0 0 1px $greyscale-60;
  color: $greyscale-40;

  &:hover {
    background-color: $greyscale-90;
  }
}

.button.primary.warning {
  background-color: $red-50;
  color: $greyscale-100;

  &.buttonLoading:hover,
  &.buttonLoading {
    background-color: $red-80;
  }

  &:hover {
    background-color: $red-60;
  }
}

.button.secondary.warning {
  background-color: $greyscale-100;
  box-shadow: inset 0 0 0 1px $red-50;
  color: $red-50;

  &.buttonLoading:hover,
  &.buttonLoading {
    background-color: $orange-80;
  }

  &:hover {
    background-color: $orange-60;
  }
}

.button.input {
  background-color: $blue-50;
  color: $greyscale-100;

  &.buttonLoading:hover,
  &.buttonLoading {
    background-color: $blue-80;
  }

  &:hover {
    background-color: $blue-60;
  }
}

.button.primary.promote {
  background-color: $orange-50;
  color: $greyscale-100;

  &.buttonLoading:hover,
  &.buttonLoading {
    background-color: $orange-80;
  }

  &:hover {
    background-color: $orange-60;
  }
}

.fullWidth {
  width: 100%;
  text-align: center;

  .text {
    float: none;
    // @include textTruncate;
  }
}

.button.secondary.promote {
  background-color: $greyscale-100;
  box-shadow: inset 0 0 0 1px $orange-50;
  color: $orange-50;

  &:hover {
    background-color: $orange-50;
    color: $greyscale-100;
  }
}

.spinner {
  margin: 0;
  position: absolute;
  left: 0;
  right: 0;
}

.spinner + .text {
  visibility: hidden;
}
