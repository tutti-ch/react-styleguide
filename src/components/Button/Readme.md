```jsx noeditor
const classes = require('./preview.scss');
<Table class="">
  <Table.Body>
    <Table.Row>
      <Table.Cell width='20%'>
        Role Warning
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} role={Button.ROLE_WARNING}>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} role={Button.ROLE_WARNING}>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} role={Button.ROLE_WARNING}>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Disabled
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} disabled>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} disabled>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} disabled>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Square borders
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} rounded={false}>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} rounded={false}>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} rounded={false}>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Size Small
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} size={Button.SIZE_SMALL}>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} size={Button.SIZE_SMALL}>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} size={Button.SIZE_SMALL}>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Size Large
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} size={Button.SIZE_LARGE}>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} size={Button.SIZE_LARGE}>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} size={Button.SIZE_LARGE}>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Loading
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} loading>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} loading>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} loading>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
    <Table.Row>
      <Table.Cell width='20%'>
        Position fullWidth
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_PRIMARY} position={Button.POSITION_FULL_WIDTH}>Primary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_SECONDARY} position={Button.POSITION_FULL_WIDTH}>Secondary</Button>
      </Table.Cell>
      <Table.Cell>
        <Button level={Button.LEVEL_TERTIARY} position={Button.POSITION_FULL_WIDTH}>Tertiary</Button>
      </Table.Cell>
    </Table.Row>
  </Table.Body>
</Table>
```