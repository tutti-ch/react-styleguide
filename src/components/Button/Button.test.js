/* global describe, test, expect */
import React from "react";
import Button from "./Button";
import { mount } from "enzyme";

describe("(Component) Button", () => {
  test("should render snapshot correctly", () => {
    const props = {
      className: "my-class"
    };

    const comp = mount(
      <Button level="primary" size="large" {...props}>
        <span className={"tutti-cube"} />
        Test content
      </Button>
    );

    expect(comp).toMatchSnapshot();
  });

  test("should be disabled if button is loading", () => {
    const props = {
      disabled: true,
      loading: true,
      type: "button",
      className: "my-class"
    };

    const comp = mount(
      <Button level="primary" size="large" {...props}>
        <span className={"tutti-cube"} />
        Test content
      </Button>
    );

    // Should be still disabled true because when state.loading is true
    // the attribute is overwritten.
    comp.setProps({ loading: true, type: "button", disabled: false });
    expect(comp).toMatchSnapshot();
  });
});
