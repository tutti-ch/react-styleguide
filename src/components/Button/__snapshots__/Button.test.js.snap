// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Button should be disabled if button is loading 1`] = `
<Button
  className="my-class"
  disabled={false}
  level="primary"
  loading={true}
  rounded={true}
  size="large"
  type="button"
>
  <button
    className="my-class"
    disabled={true}
    type="button"
  >
    <Spinner
      color="light"
      key="spinner"
      size={1.25}
    >
      <div
        className=""
      >
        <div
          style={
            Object {
              "borderBottom": "0.125rem solid rgba(255, 255, 255, 0.6)",
              "borderLeft": "0.125rem solid rgba(255, 255, 255, 0.8)",
              "borderRight": "0.125rem solid rgba(255, 255, 255, 0.6)",
              "borderTop": "0.125rem solid rgba(255, 255, 255, 0.6)",
              "height": "1.25rem",
              "width": "1.25rem",
            }
          }
        />
      </div>
    </Spinner>
    <div
      key="text"
    >
      <span
        className="tutti-cube"
      />
      Test content
    </div>
  </button>
</Button>
`;

exports[`(Component) Button should render snapshot correctly 1`] = `
<Button
  className="my-class"
  level="primary"
  loading={false}
  rounded={true}
  size="large"
  type="submit"
>
  <button
    className="my-class"
    type="submit"
  >
    <span
      className="tutti-cube"
    />
    Test content
  </button>
</Button>
`;
